fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneTypeConverter.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * LuceneTypeConverter.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene;
+
+import com.apple.foundationdb.record.metadata.IndexTypes;
+import com.apple.foundationdb.record.metadata.expressions.FieldKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.GroupingKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpressionVisitor;
+import com.apple.foundationdb.record.metadata.expressions.NestingKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.ThenKeyExpression;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Converts from normal KeyExpression types to Lucene KE types.
+ */
+public class LuceneTypeConverter implements KeyExpressionVisitor {
+
+    boolean fanNext = false;
+
+    @Override
+    public FieldKeyExpression visitField(final FieldKeyExpression fke) {
+        fanNext = false;
+        if (fke instanceof LuceneFieldKeyExpression) {
+            return fke;
+        }
+        return new LuceneFieldKeyExpression(fke, LuceneKeyExpression.FieldType.STRING, true, true);
+    }
+
+    @Override
+    public KeyExpression visitThen(final ThenKeyExpression thenKey) {
+        boolean hasPrimaryExpression = fanNext;
+        List<KeyExpression> children = thenKey.getChildren().stream()
+                .map(this::visit).collect(Collectors.toList());
+        if (!hasPrimaryExpression) {
+            return new LuceneThenKeyExpression(null, children);
+        }
+        fanNext = false;
+        return new LuceneThenKeyExpression(children.get(0), children);
```

Comment:
Should be a comment here that (for now) we assume that the first child field is a (unique or nearly so) key among the child records. In the future, we will presumably need a way to control this, perhaps as an alternative to storing wide child records as block child documents.